File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        prefixSum = defaultdict(int)
        prefixSum[0] = 1
        currSum = 0
        count = 0
        
        for i in nums:
            currSum += i
            if (currSum - k) in prefixSum:
                count += prefixSum[(currSum - k)]
            prefixSum[currSum] += 1
        
        return count
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
#         prefix=defaultdict()
#         preSum=count=0
#         prefix[0]=1
#         for i in nums:
#             preSum+=i
#             if preSum-k in prefix:
#                 count+=prefix[preSum-k]
#             if preSum in prefix:
#                 prefix[preSum]+=1
#             else:
#                 prefix[preSum]=1
        
#         return count

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int subarraySum(int[] nums, int k) {
        // Prefix Sum
        Map<Integer, Integer> prefix = new HashMap<>();
        prefix.put(0, 1); // Initialize with sum 0 occurring once
        int curSum = 0, count = 0;
        
        for (int i = 0; i < nums.length; i++) {
            curSum += nums[i];
            int val = curSum - k;
            if (prefix.containsKey(val)) {
                count += prefix.get(val);
            }
            prefix.put(curSum, prefix.getOrDefault(curSum, 0) + 1);
        }
        
        return count;
    }
}
